src/site/topics/objects-review/equals.rst

```diff
@@ -1,56 +1,167 @@
-****************
-Aside --- Equals
-****************
+****************************
+Aside --- Equals & Hash Code
+****************************
+
+Equals
+======
+
+* Below is an improved and more correct implementation of the ``equals`` method
+* You will notice the type of the parameter has changed from ``Friend`` to ``Object``
 
-* Below is a better and more correct implementation of the ``equals`` method
-* You will notice the type of the parameter has changed from ``Friend`` too ``Object``
     * The ``Object`` class is at the top of the class hierarchy
     * Every class we write will all have ``Object`` as a superclass
     * Every class we can have will actually be an ``Object``
-    * The more details on this will come later in the course
+    * More details on this will come later in the course
+
+
+* Here it checks if the ``Object`` ``o`` is the actual thing being compared to --- ``if (o == this)``
 
-* Here you will see that we check if the ``Object o`` is the actual thing we are comparing to
-    * Then obviously they are equal
+    * If they are aliases for the same object
+    * If they are, then obviously they are equal
+
+
+* It also check if it is null --- ``if (o == null)``
 
-* We also check if it is null
     * If it is null, then clearly ``this`` cannot be equal to ``o``
 
-* We check if they are of the same class or not
-    * If they are not, we say they are not equal
 
-* We then need to *downcast* the ``Object`` to something of class ``Friend``
-    * This lets us access the fields and methods for ``Friend`` from ``o``
+* It then checks if they are of the same class --- ``if (o.getClass() != this.getClass())``
+
+    * If they are not, then they are not equal
+
+
+* If the method gets to this point, it knows that ``o`` is of class ``Friend``
+* It *downcast* the ``Object`` to class ``Friend``
+
+    * This allows for accessing the fields and methods from ``Friend``
     * Again, more on this later in the course
 
-* Lastly, we check if the ``firstName`` and ``lastName`` are the same
+
+* Lastly, it check if all the attributes are equal
+
+.. note::
+
+    It is important to understand the difference between ``someObject == someOtherObject``,
+    ``someObject.equals(someOtherObject)``, and ``Objects.equals(someObject, someOtherObject)``.
+
+    With ``someObject == someOtherObject``, it checks if ``someObject`` and ``someOtherObject`` are referencing the same
+    object -- aliases.
+
+    ``someObject.equals(someOtherObject)`` checks if ``someObject`` and ``someOtherObject`` are equivalent based on
+    ``someObject`` class' ``equals`` method.
+
+    ``Objects.equals(someObject, someOtherObject)`` also checks if ``someObject`` is ``null``, and then checks if
```

Comment:
I'd maybe write this as:
> `Objects.equals(a, b)` is basically the same as `a.equals(b)`, but additionally checks if both `a` and `b` are `null` in which case they are also equal. 

Maybe worth linking the javadoc as well. Just make sure to link to a recent long term support one, like jdk17.

Reply:
I believe `Objects.equals(a, b)` only checks if `a` is null, it's `a` class' `equals` that checks `b`. 

Nevertheless, that is preferred wording.

Reply:
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Objects.html#equals(java.lang.Object,java.lang.Object)

Reply:
but it doesn't _check_ if it's null, that's just the sameness check first: 

```java
    public static boolean equals(Object a, Object b) {
        return (a == b) || (a != null && a.equals(b));
    }
    ```

Reply:
The code is irrelevant. 

> Returns true if the arguments are equal to each other and false otherwise. **Consequently, if both arguments are null, true is returned** and if exactly one argument is null, false is returned. Otherwise, equality is determined by using the [equals](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Object.html#equals(java.lang.Object)) method of the first argument.

Emphasis mine.

Reply:
Point is, it doesn't just check if `a` is `null`, it performs a null-safe equality test. The implementation is irrelevant.

Reply:
That's the `Object` class' equals, not the `Objects` class

> Returns true if the arguments are equal to each other and false otherwise. Consequently, if both arguments are null, true is returned. Otherwise, if the first argument is not null, equality is determined by calling the equals method of the first argument with the second argument of this method. Otherwise, false is returned.

Thus, null check doesn't directly happen, it only is from the sameness check. `b` is never checked for null.

Reply:
> That's the `Object` class' equals, not the `Objects` class

That static function is the `equals` method? :coffee: time, brother.